daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala

```diff
@@ -522,4 +545,22 @@ object Engine {
     StableConfig.copy(allowedLanguageVersions = LanguageVersion.DevVersions)
   )
 
+  private[engine] case class DisclosureTable(
+      allDisclosures: ImmArray[DisclosedContract],
+      contractIdByKey: Map[crypto.Hash, ContractId],
+      contractById: Map[ContractId, VersionedContractInstance],
+  )
+
+  private[engine] def buildDiscTable(disclosures: ImmArray[DisclosedContract]): DisclosureTable = {
```

Comment:
I’d suggest to catch duplicates here either way even if we check for them in the participant. The difference is mainly whether they become internal errors or expected input validation errors.

Reply:
This check is probably better done during typechecking/normalization. I add it to the ticket.

Reply:
The PoC participant currently doesn't check for duplicate contract ids/keys as part of the GRPC command validation. Is that something the participant should do or can we rely on the engine to catch all invalid inputs?

Reply:
Doing it all in the engine sounds easier so I’d go with that unless anyone has good arguments to go the other way.